design/recovery-internals.md

```diff
@@ -0,0 +1,142 @@
+
+
+Proxies haven’t recovered the transaction system states and cannot accept transactions yet. The master recovers proxies’ states by sending the txnStateStore to proxies through proxies’ (txnState) interfaces in sendIntialCommitToResolvers() function. Once proxies have recovered their states, they can start processing transactions. The recovery transaction that was waiting on proxies will be processed.
+
+
+The resolvers haven’t known the recovery version either. The master needs to send the lastEpochEnd version (i.e., last commit of the previous generation) to resolvers via resolvers’ (resolve) interface.
+
+
+At the end of this phase, every role should be aware of the recovery and start recovering its states.
+
+
+## Phase 5: WRITING_CSTATE
+
+Coordinators stores the transaction systems’ information. The master needs to write the new tLogs into coordinators’ states to achieve consensus and fault tolerance. Only when the coordinators’ state is updated with the new transaction system’s configuration, will the cluster controller tell clients about the new transaction system (such as the new proxies).
+
+
+Once the cstate is written, the master sets the cstateUpdated promise and move to the ACCEPTING_COMMITS phase.
+
+
+The cstate update is done in trackTlogRecovery() actor.
+The actor keeps running until the recovery finishes the FULLY_RECOVERED phase. 
+The actor needs to update the cstates at the following phases:
+ALL_LOGS_RECRUITED, STORAGE_RECOVERED, and FULLY_RECOVERED. 
+For example, when the old tLogs are no longer needed, the master will write the coordinators’ state again.
+
+
+Now the main steps in recovery have finished. The master keeps waiting for all tLogs to join the system and for all storage servers to roll back their prefetched data before claiming the system is fully recovered. 
+
+
+## Phase 6: ACCEPTING_COMMITS
+
+The transaction system starts to accept new transactions.
+
+
+## Phase 7: ALL_LOGS_RECRUITED
+
+The master marks the recovery phase to ALL_LOGS_RECRUITED when the number of new tLogs it receives is equal to the configured expected tLogs. This is done in the trackTlogRecovery() actor.
+
+
+## Phase 8: STORAGE_RECOVERED
+
+Storage servers need old tLogs in previous generations to recover storage servers’ state. For example, a storage server may be offline for a long time, lagging behind in pulling mutations assigned to it. We have to keep the old tLogs who have those mutations until no storage server needs them. 
+
+When all tLogs are no longer needed and deleted, the master move to the STORAGE_RECOVERED phase. This is done by checking if oldTLogData is empty in the trackTlogRecovery() actor.
```

Comment:
```suggestion
When all tLogs are no longer needed and deleted, the master move to the `STORAGE_RECOVERED` phase. This is done by checking if `oldTLogData` is empty in the `trackTlogRecovery()` actor.
```

Reply:
I didn't wrap the phase name, say STORAGE_RECOVERED, into "``", because I think those names have been capitalized and easier to distinguish from other texts.

To keep this consistent in the doc, I won't add `` here unless you prefer adding `` to all places that have the capitalized phase name or the writing style (say Chicago style) has some rules about this.